lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,226 @@
+        # on-disk license file (which licensee may not find due to a non-standard filename)
+        def nuspec_local_license_file
+          return @nuspec_local_license_file if defined?(@nuspec_local_license_file)
+          return unless nuspec_contents
+
+          match = @nuspec_contents.match LICENSE_FILE_REGEX
+          if match && match[1]
+            license_path = File.join(File.dirname(nuspec_path), match[1])
+            return unless File.exist?(license_path)
+            license_data = File.read(license_path)
+            @nuspec_local_license_file = Licensee::ProjectFiles::LicenseFile.new(license_data, license_path)
+          end
+        end
+
+        # Look for a <licenseUrl> element in the nuspec that either is known to contain a license identifier
+        # in the URL, or points to license text on the internet that can be downloaded.
+        def nuspec_remote_license_file
+          return @nuspec_remote_license_file if defined?(@nuspec_remote_license_file)
+          return unless nuspec_contents
+
+          match = nuspec_contents.match LICENSE_URL_REGEX
+          if match && match[1]
+            url = original_url = match[1]
+
+            # Skip downloading if the URL contains a license identifier itself
+            if known_license = inspect_url_for_license(url)
+              return @nuspec_remote_license_file = KnownLicenseFile.new(known_license, { uri: url })
+            end
+
+            unless ignored_url?(url)
+              # Transform URLs that are known to return HTML but have a corresponding text-based URL
+              url = get_text_content_url(url)
+
+              # Attempt to fetch the license content
+              license_data = retrieve_license(url)
+              unless license_data.nil?
+                @nuspec_remote_license_file = Licensee::ProjectFiles::LicenseFile.new(license_data, { uri: original_url })
+              end
+            end
+          end
+        end
+
+        def get_text_content_url(url)
+          # Convert github file URLs to raw URLs
+          if match = url.match(/https?:\/\/(?:www\.)?github.com\/([^\/]+)\/([^\/]+)\/blob\/(.*)/i)
+            url = "https://github.com/#{match[1]}/#{match[2]}/raw/#{match[3]}"
+          else
+            url
+          end
+        end
+
+        def inspect_url_for_license(url)
+          if match = url.match(/https?:\/\/licenses.nuget.org\/(.*)/i)
+            Licensee::License.find(match[1])
+          elsif match = url.match(/https?:\/\/opensource.org\/licenses\/(.*)/i)
+            Licensee::License.find(match[1])
+          elsif match = url.match(/https?:\/\/(?:www\.)?apache.org\/licenses\/(.*?)(?:\.html|\.txt)?$/i)
+            Licensee::License.find(match[1].gsub("LICENSE", "Apache"))
+          end
+        end
+
+        def ignored_url?(url)
+          url == "https://aka.ms/deprecateLicenseUrl"
+        end
+
+        def retrieve_license(url, redirect_limit = 2)
```

Comment:
what are the chances that two dependencies point at the same license url here?  just wondering if the output from this function should be cached/memoized

Reply:
Great point - that's in fact very common and definitely worth doing.